api/logfilter.go

```diff
@@ -0,0 +1,109 @@
+package api
+
+import (
+	"bytes"
+
+	"github.com/iotexproject/iotex-proto/golang/iotexapi"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
+	"go.uber.org/zap"
+
+	"github.com/iotexproject/iotex-core/blockchain/block"
+	"github.com/iotexproject/iotex-core/pkg/log"
+)
+
+// LogFilter contains options for contract log filtering.
+type LogFilter struct {
+	stream  iotexapi.APIService_StreamFilterLogsServer
+	errChan chan error
+	*iotexapi.FilterLogsRequest
+	// FilterLogsRequest.Topics restricts matches to particular event topics. Each event has a list
+	// of topics. Topics matches a prefix of that list. An empty element slice matches any
+	// topic. Non-empty elements represent an alternative that matches any of the
+	// contained topics.
+	//
+	// Examples:
+	// {} or nil          matches any topic list
+	// {{A}}              matches topic A in first position
+	// {{}, {B}}          matches any topic in first position, B in second position
+	// {{A}, {B}}         matches topic A in first position, B in second position
+	// {{A, B}}, {C, D}}  matches topic (A OR B) in first position, (C OR D) in second position
+}
+
+// NewLogFilter returns a new log filter
+func NewLogFilter(in *iotexapi.FilterLogsRequest, stream iotexapi.APIService_StreamFilterLogsServer, errChan chan error) Responder {
+	return &LogFilter{
+		FilterLogsRequest: in,
+		stream:            stream,
+		errChan:           errChan,
+	}
+}
+
+// Respond to new block
+func (l *LogFilter) Respond(blk *block.Block) error {
+	logs := l.matchBlock(blk)
+	if len(logs) == 0 {
+		return nil
+	}
+	// send matched logs thru streaming API
+	for _, e := range logs {
+		err := l.stream.Send(e)
+		if err != nil {
+			l.errChan <- err
+			log.L().Info("error streaming the log",
+				zap.Uint64("height", e.BlkHeight),
+				zap.Error(err))
+			return err
+		}
+	}
+	return nil
+}
+
+// Exit send to error channel
+func (l *LogFilter) Exit() {
+	l.errChan <- nil
+}
+
+// matchBlock returns matching logs in a given block
+func (l *LogFilter) matchBlock(blk *block.Block) []*iotextypes.Log {
+	var logs []*iotextypes.Log
+	for _, r := range blk.Receipts {
+		for _, v := range r.Logs {
+			log := v.ConvertToLogPb()
+			if l.match(log) {
+				logs = append(logs, log)
+			}
+		}
+	}
+	return logs
+}
+
+// match checks if a given log matches the filter
+func (l *LogFilter) match(log *iotextypes.Log) bool {
+	if l.Address != log.ContractAddress {
+		return false
+	}
+	if len(l.Topics) > len(log.Topics) {
```

Comment:
L85-L107 could be simplified into the logic that if there exists at least one topic in log.Topics that doesn't match a one in l.Topics (note that missing isn't considered as mismatch)